File: backend-node/services/extractor.js
const { PDFParse } = require('pdf-parse');

// =====================================================================
// PO EXTRACTOR - pure code, regex-based extraction (no AI / no API key)
//
// This module reads raw text from a PDF (via pdf-parse v2) and extracts
// observable facts using pattern matching. Three layouts are supported,
// matching the three customer formats in the sample documents:
//
//   - BrightKids  : simple flat table layout
//   - ThreadHaven : multi-page delivery splits / prepacks (George POs)
//   - Raj International Sourcing : ratio-pack table layout
//
// The output shape (extracted_facts) matches what transformer.js expects,
// so transformer.js and validator.js are UNCHANGED from the rest of the
// pipeline - only this extraction step differs.
// =====================================================================

// =====================================================================
// LAYOUT 1 - BrightKids
// =====================================================================
function extractBrightKids(text) {
  const documentBrand = text.trim().split('\n')[0].trim();
  const poNumber = (text.match(/Purchase Order Number\s*(\d+)/) || [])[1] || '';
  const vendorNo = (text.match(/\nVendor\s*(\d+)/) || [])[1] || '';
  const factoryNo = (text.match(/\nFactory\s*(\d+)/) || [])[1] || '';
  const productDescription = (text.match(/Product Description\s*(.+?)\n/) || [])[1]?.trim() || '';
  const colour = (text.match(/Colour\s*(.+?)\n/) || [])[1]?.trim() || '';
  const season = (text.match(/Season\s+([A-Z]{2}\d{2})/) || [])[1] || '';
  const styleNo = (text.match(/Product Ident\s*([\d-]+)/) || [])[1] || '';
  const currency = (text.match(/Currency Code\s*([A-Z]{3})/) || [])[1] || '';
  const countryOfOrigin = (text.match(/Country of Origin\s*(.+?)\n/) || [])[1]?.trim() || '';
  const portOfDeparture = (text.match(/Port of Departure\s*(.+?)\n/) || [])[1]?.trim() || '';
  const paymentTerms = (text.match(/Payment Terms\s*(Net Due \d+ days)/) || [])[1] || '';

  // Each delivery: "Delivery N Shipment Date: DD-MMM-YY" followed by a
  // size table, ending at "TOTALS"
  const deliveryRegex = /Delivery (\d+) Shipment Date:\s*([A-Za-z0-9-]+)\nSize Barcode Cost \(USD\) Retail \(USD\) Order Qty Cartons\n([\s\S]*?)TOTALS/g;
  // Item rows are now cleanly space-separated: size barcode cost retail order_qty cartons
  const itemRowRegex = /(\d{1,2}-\d{1,2}Y) (\d{13}) (\d+\.\d{2}) (\d+\.\d{2}) (\d+) (\d+)/g;

  const deliveries = [];
  let m;
  while ((m = deliveryRegex.exec(text)) !== null) {
    const [, deliveryNum, shipmentDate, block] = m;
    const items = [];
    let im;
    itemRowRegex.lastIndex = 0;
    while ((im = itemRowRegex.exec(block)) !== null) {
      const [, size, barcode, cost, retail, orderQty, cartons] = im;
      items.push({ size, barcode, cost_price: cost, retail_price: retail, order_qty: orderQty, cartons });
    }
    deliveries.push({ delivery_number: deliveryNum, shipment_date: shipmentDate, items });
  }

  return {
    document_brand: documentBrand,
    po_number: poNumber,
    vendor_no: vendorNo,
    supplier_no: '',
    factory_no: factoryNo,
    style_no: styleNo,
    product_description: productDescription,
    season: season,
    colour: colour,
    currency: currency,
    country_of_origin: countryOfOrigin,
    port_of_departure: portOfDeparture,
    lading_port: '',
    destination_location: '',
    order_date: '',
    payment_terms: paymentTerms,
    incoterm: '',
    items: [],
    deliveries: deliveries,
    prepacks: [],
    ratio_packs: [],
    missing_fields: []
  };
}

// =====================================================================
// LAYOUT 2 - ThreadHaven / George (delivery splits + prepacks)
// =====================================================================
function extractGeorge(text) {
  const documentBrand = 'ThreadHaven';
  const poNumber = (text.match(/Purchase Order Number\s*(\d+)/) || [])[1] || '';
  const vendorNo = (text.match(/\nVendor\s*(\d+)/) || [])[1] || '';
  const factoryNo = (text.match(/\nFactory\s*(\d+)/) || [])[1] || '';
  const productDescription = (text.match(/Product Description\s*(.+?)\s*QC Status/) || [])[1]?.trim() || '';
  const colour = (text.match(/Colour\s*(.+?)\s*Launch Date/) || [])[1]?.trim() || '';
  const season = (text.match(/\nSeason\s+([A-Z]{2}\d{2})/) || [])[1] || '';
  const styleNo = (text.match(/Product Ident\s*([\d-]+)/) || [])[1] || '';
  const currency = (text.match(/Currency Code\s*([A-Z]{3})/) || [])[1] || '';
  const countryOfOrigin = (text.match(/Country of Origin\s*(.+?)\s*Refurb/) || [])[1]?.trim() || '';
  const portOfDeparture = (text.match(/Port of Departure\s*(.+?)\s*Refurb Cost/) || [])[1]?.trim() || '';
  const paymentTerms = (text.match(/Payment Terms\s*(.+?)\s*Pack Type/) || [])[1]?.trim() || '';

  // Detect delivery numbers + shipment dates (dedupe across page breaks)
  const deliveryMatches = [...text.matchAll(/Delivery (\d+) Split A\nShipment Date:\s*([A-Za-z0-9 ]+?)\s*-\n/g)];
  const seenDeliveries = new Map();
  for (const dm of deliveryMatches) {
    if (!seenDeliveries.has(dm[1])) {
      seenDeliveries.set(dm[1], dm[2].trim());
    }
  }

  // Item rows: SizeLabel + 4 numeric columns (variant/win/sizecode/barcode)
  // + cost price + retail price. Order/cartons not needed for items[].
  const itemRowRegex = /(\d{1,2}-\d{1,2}Y|10PC PRE-\nPACK)(?:\s+\d+){4}\s+(\d+\.\d{2})\s+(\d+\.\d{2})/g;
  const allItemsRaw = [...text.matchAll(itemRowRegex)].map(m => ({
    size: m[1].replace(/\n/g, ' ').trim(),
    cost_price: m[2],
    retail_price: m[3]
  }));

  // Separate the prepack "container" row (if any) from regular size items
  const isPrepack = text.includes('PRE-\nPACK') || text.includes('PREPACK');
  const allItems = allItemsRaw.filter(it => !it.size.includes('PRE'));

  // Build deliveries array, distributing items evenly across delivery splits
  const deliveryNums = [...seenDeliveries.keys()];
  const deliveries = [];

  if (deliveryNums.length > 1 && allItems.length % deliveryNums.length === 0) {
    // Multiple delivery splits with the same item set repeated per split
    const perDelivery = allItems.length / deliveryNums.length;
    deliveryNums.forEach((num, i) => {
      deliveries.push({
        delivery_number: num,
        shipment_date: seenDeliveries.get(num),
        items: allItems.slice(i * perDelivery, (i + 1) * perDelivery)
      });
    });
  } else {
    // Single delivery (e.g. prepack PO) - all items in one delivery
    const num = deliveryNums[0] || '1';
    deliveries.push({
      delivery_number: num,
      shipment_date: seenDeliveries.get(num) || '',
      items: allItems
    });
  }

  // Prepack detection - build a prepacks entry from the size list
  let prepacks = [];
  if (isPrepack) {
    prepacks = [{
      prepack_id: '1',
      sizes: allItems.map(it => ({ size: it.size }))
    }];
  }

  return {
    document_brand: documentBrand,
    po_number: poNumber,
    vendor_no: vendorNo,
    supplier_no: '',
    factory_no: factoryNo,
    style_no: styleNo,
    product_description: productDescription,
    season: season,
    colour: colour,
    currency: currency,
    country_of_origin: countryOfOrigin,
    port_of_departure: portOfDeparture,
    lading_port: '',
    destination_location: '',
    order_date: '',
    payment_terms: paymentTerms,
    incoterm: '',
    items: [],
    deliveries: deliveries,
    prepacks: prepacks,
    ratio_packs: [],
    missing_fields: []
  };
}

// =====================================================================
// LAYOUT 3 - Raj International Sourcing (ratio pack table)
// =====================================================================
function extractRaj(text) {
  const documentBrand = 'Raj International Sourcing';
  const poNumber = (text.match(/\b(\d{3}-\d{5})\b/) || [])[1] || '';
  const supplierNo = (text.match(/Supplier No\.\s*:\s*(\d+)/) || [])[1] || '';
  const destinationLocation = (text.match(/Destination Location\s*:\s*(.+?)\n/) || [])[1]?.trim() || '';
  const orderDate = (text.match(/Order Create Date\s*:\s*([A-Za-z0-9-]+)/) || [])[1] || '';
  const styleNo = (text.match(/\n([A-Z]{2}\d{6})\n/) || [])[1] || '';
  const incoterm = text.includes('FOB') ? 'FOB' : '';

  // Country of origin: "Bangladesh" appears near the top of the document
  // as the sourcing origin (Raj ships ex-Chittagong, Bangladesh)
  const countryOfOrigin = /Bangladesh\nWomens/.test(text) ? 'Bangladesh' : '';

  // Season code: printed as standalone "AW" (without year suffix)
  const seasonRaw = /\nAW\n/.test(text) ? 'AW' : '';

  // Ratio pack table: colours, sizes and unit costs appear as parallel
  // duplicated arrays in the raw text (pdf-parse extracts table cells in
  // column order, not visual reading order). We extract the 5 sizes and
  // the unit cost (USD per-unit price, excluding ratio-pack-level totals).
  const colourBlock = (text.match(/(?:BLACK\n?)+/) || [])[0] || '';
  const hasColour = colourBlock.includes('BLACK');

  const sizeBlockMatch = text.match(/(?:BLACK\n){5}([\d\n]+?)(?=\d{10,})/);
  const sizes = sizeBlockMatch
    ? sizeBlockMatch[1].split('\n').filter(Boolean)
    : [];

  // Unit cost - first repeated "USD X.XX" value (per-unit cost)
  const unitCostMatch = (text.match(/USD (\d+\.\d{2})/) || [])[1] || '';

  const ratioPacks = sizes.length > 0
    ? [{
        pack_name: 'RATIO PACK',
        sizes: sizes.map(s => ({ size: s, cost_price: unitCostMatch }))
      }]
    : [];

  // Items: one per size variant, using the unit cost for price
  const items = sizes.map(s => ({
    size: s,
    cost_price: unitCostMatch
  }));

  return {
    document_brand: documentBrand,
    po_number: poNumber,
    vendor_no: '',
    supplier_no: supplierNo,
    factory_no: '',
    style_no: styleNo,
    product_description: '',
    season: seasonRaw,
    colour: hasColour ? 'BLACK' : '',
    currency: 'USD',
    country_of_origin: countryOfOrigin,
    port_of_departure: 'Chittagong - Bangladesh',
    lading_port: 'Chittagong - Bangladesh',
    destination_location: destinationLocation,
    order_date: orderDate,
    payment_terms: '',
    incoterm: incoterm,
    items: items,
    deliveries: [],
    prepacks: [],
    ratio_packs: ratioPacks,
    missing_fields: []
  };
}

// =====================================================================
// LAYOUT DETECTION
// =====================================================================
function detectLayout(text) {
  if (text.includes('BrightKids')) return 'brightkids';
  if (text.includes('International Sourcing') || text.includes('Raj Limited')) return 'raj';
  if (text.includes('Purchase Order Number') && text.includes('Split A')) return 'george';
  return 'unknown';
}

// =====================================================================
// MAIN EXTRACTION FUNCTION
// =====================================================================
async function extractFromPDF(pdfBase64, filename) {
  try {
    const buffer = Buffer.from(pdfBase64, 'base64');
    const parser = new PDFParse({ data: buffer });
    const result = await parser.getText();
    const text = result.text;

    const layout = detectLayout(text);
    let facts;

    switch (layout) {
      case 'brightkids':
        facts = extractBrightKids(text);
        break;
      case 'george':
        facts = extractGeorge(text);
        break;
      case 'raj':
        facts = extractRaj(text);
        break;
      default:
        return {
          success: false,
          filename: filename,
          error: `Could not detect a known document layout for ${filename}`,
          extracted_facts: null
        };
    }

    return {
      success: true,
      filename: filename,
      extracted_facts: facts
    };
  } catch (error) {
    return {
      success: false,
      filename: filename,
      error: error.message,
      extracted_facts: null
    };
  }
}

module.exports = { extractFromPDF, detectLayout, extractBrightKids, extractGeorge, extractRaj };

File: backend-python/services/extractor.py
"""
PO EXTRACTOR - pure code, pdfplumber-based extraction (no AI / no API key)

Reads a PDF using pdfplumber and extracts observable facts using a mix of
table extraction and pattern matching. Three layouts are supported, matching
the three customer formats in the sample documents:

  - BrightKids  : simple flat table, items extracted from a clean pdfplumber
                  table (Size/Barcode/Cost/Retail/Order Qty/Cartons)
  - ThreadHaven : multi-page delivery splits / prepacks (George POs).
                  Header fields come from a clean key-value table; items are
                  extracted via regex from the page text (pdfplumber's table
                  detection is unreliable for this item table layout).
  - Raj International Sourcing : ratio-pack table layout. Header fields are
                  clean "Label : Value" pairs in the page text; items come
                  from the Colour/Size/.../Unit cost tables that pdfplumber
                  extracts cleanly for this layout.

The output shape (extracted_facts) matches the original AI-based extractor's
output, so transformer.py and validator.py do not need any layout-specific
knowledge.
"""

import io
import re
import pdfplumber


def _fmt_price(value):
    """Format a price string like '9.6200' -> '9.62'."""
    try:
        return f'{float(value):.2f}'
    except (TypeError, ValueError):
        return value or ''


# =====================================================================
# LAYOUT 1 - BrightKids
# =====================================================================
def extract_brightkids(page):
    text = page.extract_text()
    tables = page.extract_tables()

    document_brand = text.strip().split('\n')[0].strip()
    po_number = re.search(r'Purchase Order Number\s*(\d+)', text).group(1)
    vendor_no = re.search(r'\nVendor\s*(\d+)', text).group(1)
    factory_no = re.search(r'\nFactory\s*(\d+)', text).group(1)
    product_description = re.search(r'Product Description\s*(.+?)\s*Currency Code', text).group(1).strip()
    colour = re.search(r'Colour\s*(.+?)\s*Port of Departure', text).group(1).strip()
    season = re.search(r'Season\s+([A-Z]{2}\d{2})', text).group(1)
    style_no = re.search(r'Product Ident\s*([\d-]+)', text).group(1)
    currency = re.search(r'Currency Code\s*([A-Z]{3})', text).group(1)
    country_of_origin = re.search(r'Country of Origin\s*(.+?)\n', text).group(1).strip()
    port_of_departure = re.search(r'Port of Departure\s*(.+?)\n', text).group(1).strip()
    payment_terms_m = re.search(r'Payment Terms\s*(Net Due \d+ days)', text)
    payment_terms = payment_terms_m.group(1) if payment_terms_m else ''

    # Items table: rows are [Size, Barcode, Cost, Retail, Order Qty, Cartons]
    # with "Delivery N Shipment Date: ..." marker rows separating deliveries
    item_table = None
    for table in tables:
        if table and len(table[0]) == 6 and table[1] == ['Size', 'Barcode', 'Cost (USD)', 'Retail (USD)', 'Order Qty', 'Cartons']:
            item_table = table
            break

    deliveries = []
    current = None
    if item_table:
        for row in item_table:
            col0 = row[0] or ''
            m = re.match(r'Delivery (\d+) Shipment Date:\s*([A-Za-z0-9-]+)', col0)
            if m:
                if current:
                    deliveries.append(current)
                current = {'delivery_number': m.group(1), 'shipment_date': m.group(2), 'items': []}
                continue
            if col0 in ('Size', 'TOTALS', ''):
                continue
            if re.match(r'\d{1,2}-\d{1,2}Y', col0) and current is not None:
                current['items'].append({
                    'size': row[0],
                    'barcode': row[1],
                    'cost_price': row[2],
                    'retail_price': row[3],
                    'order_qty': row[4],
                    'cartons': row[5]
                })
        if current:
            deliveries.append(current)

    return {
        'document_brand': document_brand,
        'po_number': po_number,
        'vendor_no': vendor_no,
        'supplier_no': '',
        'factory_no': factory_no,
        'style_no': style_no,
        'product_description': product_description,
        'season': season,
        'colour': colour,
        'currency': currency,
        'country_of_origin': country_of_origin,
        'port_of_departure': port_of_departure,
        'lading_port': '',
        'destination_location': '',
        'order_date': '',
        'payment_terms': payment_terms,
        'incoterm': '',
        'items': [],
        'deliveries': deliveries,
        'prepacks': [],
        'ratio_packs': [],
        'missing_fields': []
    }


# =====================================================================
# LAYOUT 2 - ThreadHaven / George (delivery splits + prepacks)
# =====================================================================
def extract_george(pdf):
    pages_text = [page.extract_text() for page in pdf.pages]
    full_text = '\n'.join(pages_text)

    # Header fields from the clean key-value table on page 1
    header_table = pdf.pages[0].extract_tables()[0]
    kv = {}
    for row in header_table:
        if row[0] and row[1]:
            kv[row[0]] = row[1]
        if len(row) > 2 and row[2] and row[3]:
            kv[row[2]] = row[3]

    document_brand = 'ThreadHaven'
    po_number = kv.get('Purchase Order Number', '')
    vendor_no = kv.get('Vendor', '')
    factory_no = kv.get('Factory', '')
    product_description = kv.get('Product Description', '')
    colour = kv.get('Colour', '')
    season = kv.get('Season', '')
    style_no = kv.get('Product Ident', '')
    currency = kv.get('Currency Code', '')
    country_of_origin = kv.get('Country of Origin', '')
    port_of_departure = kv.get('Port of Departure', '')
    payment_terms = kv.get('Payment Terms', '')

    # Vendor/Factory live in a second row of the header table on page 1,
    # but to be safe also try the page text directly
    if not vendor_no:
        m = re.search(r'\nVendor\s*(\d+)', full_text)
        vendor_no = m.group(1) if m else ''
    if not factory_no:
        m = re.search(r'\nFactory\s*(\d+)', full_text)
        factory_no = m.group(1) if m else ''

    # Delivery markers (dedupe across page breaks)
    delivery_matches = re.findall(r'Delivery (\d+) Split A\nShipment Date:\s*([A-Za-z0-9 ]+?)\s*-\n', full_text)
    seen_deliveries = {}
    for num, date in delivery_matches:
        if num not in seen_deliveries:
            seen_deliveries[num] = date.strip()

    # Item rows from text: Size + 4 numeric columns + cost + retail.
    # Note: pdfplumber renders the prepack cell's two lines as "10PC PRE-"
    # immediately followed by the row's numeric columns, with "PACK" pushed
    # to a separate line after the whole row - so we match "10PC PRE-" alone.
    item_row_regex = re.compile(r'(\d{1,2}-\d{1,2}Y|10PC PRE-)(?:\s+\d+){4}\s+(\d+\.\d{2})\s+(\d+\.\d{2})')
    all_items_raw = [
        {'size': m.group(1).replace('\n', ' ').strip(), 'cost_price': m.group(2), 'retail_price': m.group(3)}
        for m in item_row_regex.finditer(full_text)
    ]

    is_prepack = '10PC PRE-' in full_text or 'PREPACK' in full_text
    all_items = [it for it in all_items_raw if 'PRE' not in it['size']]

    delivery_nums = list(seen_deliveries.keys())
    deliveries = []

    if len(delivery_nums) > 1 and len(all_items) % len(delivery_nums) == 0:
        per_delivery = len(all_items) // len(delivery_nums)
        for i, num in enumerate(delivery_nums):
            deliveries.append({
                'delivery_number': num,
                'shipment_date': seen_deliveries[num],
                'items': all_items[i * per_delivery:(i + 1) * per_delivery]
            })
    else:
        num = delivery_nums[0] if delivery_nums else '1'
        deliveries.append({
            'delivery_number': num,
            'shipment_date': seen_deliveries.get(num, ''),
            'items': all_items
        })

    prepacks = []
    if is_prepack:
        prepacks = [{
            'prepack_id': '1',
            'sizes': [{'size': it['size']} for it in all_items]
        }]

    return {
        'document_brand': document_brand,
        'po_number': po_number,
        'vendor_no': vendor_no,
        'supplier_no': '',
        'factory_no': factory_no,
        'style_no': style_no,
        'product_description': product_description,
        'season': season,
        'colour': colour,
        'currency': currency,
        'country_of_origin': country_of_origin,
        'port_of_departure': port_of_departure,
        'lading_port': '',
        'destination_location': '',
        'order_date': '',
        'payment_terms': payment_terms,
        'incoterm': '',
        'items': [],
        'deliveries': deliveries,
        'prepacks': prepacks,
        'ratio_packs': [],
        'missing_fields': []
    }


# =====================================================================
# LAYOUT 3 - Raj International Sourcing (ratio pack table)
# =====================================================================
def extract_raj(pdf):
    pages_text = [page.extract_text() for page in pdf.pages]
    full_text = '\n'.join(pages_text)

    all_tables = []
    for page in pdf.pages:
        all_tables.extend(page.extract_tables())

    document_brand = 'Raj International Sourcing'

    po_match = re.search(r'Purchase Order\s+(\d{3}-\d{5})', full_text) or \
        re.search(r'Purchase Order\s*:\s*(\d{3}-\d{5})', full_text)
    po_number = po_match.group(1) if po_match else ''

    supplier_match = re.search(r'Supplier No\.\s*:\s*(\d+)', full_text)
    supplier_no = supplier_match.group(1) if supplier_match else ''

    dest_match = re.search(r'Destination Location\s*:\s*(.+)', full_text)
    destination_location = dest_match.group(1).strip() if dest_match else ''

    country_match = re.search(r'Country of Origin\s*:\s*(.+)', full_text)
    country_of_origin = country_match.group(1).strip() if country_match else ''

    lading_match = re.search(r'Lading Port\s*:\s*(.+)', full_text)
    lading_port = lading_match.group(1).strip() if lading_match else ''

    incoterm_match = re.search(r'Purchase Terms\s*:\s*(\w+)', full_text)
    incoterm = incoterm_match.group(1) if incoterm_match else ''

    order_date_match = re.search(r'Order Create Date\s*:\s*([A-Za-z0-9-]+)', full_text)
    order_date = order_date_match.group(1) if order_date_match else ''

    # Style No / Season Code / Product Description / unit cost - found inside
    # a 4x3 "VPN / Style No / 24 Char Description / Brand" table per ratio pack
    style_no = ''
    season = ''
    product_description = ''

    for table in all_tables:
        for row in table:
            for cell in (row or []):
                if not cell:
                    continue
                if not style_no and 'Style No' in cell:
                    m = re.search(r'Style No\s*:\s*(\S+)', cell)
                    if m:
                        style_no = m.group(1)
                if not season and 'Season Code' in cell:
                    m = re.search(r'Season Code\s*:\s*(\w+)', cell)
                    if m:
                        season = m.group(1)
                if not product_description and 'Product Description' in cell:
                    m = re.search(r'Product Description\s*:\s*(.+?)(?:\s+Average|$)', cell)
                    if m:
                        product_description = m.group(1).strip()

    # Items: Colour/Size/.../Unit cost tables (one per ratio pack)
    items = []
    for table in all_tables:
        header = table[0] if table else None
        if header and header[0] == 'Colour' and len(header) > 1 and header[1] == 'Size':
            unit_cost_idx = header.index('Unit cost') if 'Unit cost' in header else None
            for row in table[1:]:
                if row[0] == 'BLACK':
                    cost_raw = row[unit_cost_idx] if unit_cost_idx is not None else ''
                    cost_match = re.search(r'USD\s*([\d.]+)', cost_raw or '')
                    items.append({
                        'size': row[1],
                        'cost_price': _fmt_price(cost_match.group(1)) if cost_match else ''
                    })

    ratio_packs = [{
        'pack_name': 'RATIO PACK',
        'sizes': [{'size': it['size'], 'cost_price': it['cost_price']} for it in items]
    }] if items else []

    return {
        'document_brand': document_brand,
        'po_number': po_number,
        'vendor_no': '',
        'supplier_no': supplier_no,
        'factory_no': '',
        'style_no': style_no,
        'product_description': product_description,
        'season': season,
        'colour': 'BLACK' if items else '',
        'currency': 'USD',
        'country_of_origin': country_of_origin,
        'port_of_departure': lading_port,
        'lading_port': lading_port,
        'destination_location': destination_location,
        'order_date': order_date,
        'payment_terms': '',
        'incoterm': incoterm,
        'items': items,
        'deliveries': [],
        'prepacks': [],
        'ratio_packs': ratio_packs,
        'missing_fields': []
    }


# =====================================================================
# LAYOUT DETECTION
# =====================================================================
def detect_layout(text):
    if 'BrightKids' in text:
        return 'brightkids'
    if 'International Sourcing' in text or 'Raj Limited' in text:
        return 'raj'
    if 'Purchase Order Number' in text and 'Split A' in text:
        return 'george'
    return 'unknown'


# =====================================================================
# MAIN EXTRACTION FUNCTION
# =====================================================================
def extract_from_pdf(pdf_bytes, filename):
    try:
        with pdfplumber.open(io.BytesIO(pdf_bytes)) as pdf:
            first_page_text = pdf.pages[0].extract_text() or ''
            full_text = '\n'.join((page.extract_text() or '') for page in pdf.pages)

            layout = detect_layout(full_text if 'Purchase Order Number' in full_text else first_page_text)
            # detect_layout needs the right text depending on layout; use full_text generally
            layout = detect_layout(full_text)

            if layout == 'brightkids':
                facts = extract_brightkids(pdf.pages[0])
            elif layout == 'george':
                facts = extract_george(pdf)
            elif layout == 'raj':
                facts = extract_raj(pdf)
            else:
                return {
                    'success': False,
                    'filename': filename,
                    'error': f'Could not detect a known document layout for {filename}',
                    'extracted_facts': None
                }

        return {
            'success': True,
            'filename': filename,
            'extracted_facts': facts
        }
    except Exception as error:
        return {
            'success': False,
            'filename': filename,
            'error': str(error),
            'extracted_facts': None
        }
